File: ruby/binary_tree_inorder_traversal.rb
# Given a binary tree, return the inorder traversal of its nodes' values.

# For example:
# Given binary tree {1,#,2,3},
#    1
#     \
#      2
#     /
#    3
# return [1,3,2].

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}

require_relative 'src/tree.rb'

class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

def inorder_traversal(root)
	stack = []
	inorder_list = []

	while root || stack.size != 0
		if root
			stack << root
			root = root.left
		else
			root = stack.pop
			inorder_list << root.val

			if root.right
				root = root.right
			else
				root = nil
			end
		end
	end

	inorder_list
end

File: javascript/binary_tree_inorder_traversal.js
// Given a binary tree, return the inorder traversal of its nodes' values.

// For example:
// Given binary tree {1,#,2,3},
//    1
//     \
//      2
//     /
//    3
// return [1,3,2].

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var treeModule = require('./src/tree.js');

var inorderTraversal = function(root) {
	if (root === null)
		return [];

	var order, stack;
	order = [];
	stack = [];

	while (root !== null || stack.length !== 0)
	{
		if (root !== null)
		{
			stack.push(root);
			root = root.left;			
		}
		else
		{
			root = stack.pop();
			order.push(root.val);
			root = root.right;
		}
	}

	return order;
};

var root = treeModule.createTree();
treeModule.preorderTraverse(root);
var order;
order = inorderTraversal(root);
console.log(order);